File: need4spd/euler_21.rb
def getDivisors(num)
    
  divisors=Array.new(0)
  startnum = 1
  
  if num==1
    divisors.push(1)
    return divisors
  end
  
  while true
    
    if num%startnum == 0
        divisors.push(startnum)
    end
    
    startnum+=1
    if num==startnum
        break
    end
  end
  return divisors
end

targetnum=10000
resultList=Array.new(1, 0)

while (true)
  if resultList.count(targetnum) > 0
     targetnum=targetnum-1
     next      
  end      
        
  divisors=getDivisors(targetnum) #targetnum 220
  sumOfDivisors=divisors.inject(:+) #sumofDivisors 284
        
  tempDivisors=getDivisors(sumOfDivisors)
  tempSumOfDivisors=tempDivisors.inject(:+)
        
  if targetnum==tempSumOfDivisors and targetnum != sumOfDivisors
     resultList.push(targetnum)
     resultList.push(sumOfDivisors)
  end
  
  targetnum=targetnum-1
        
  if targetnum==0
      break
  end
end

puts "result #{resultList}"
puts resultList.inject(:+)

File: need4spd/euler_21.pl
use strict;
use warnings;
use List::Util qw(first sum);

sub getDivisors {
    
    my ($num) = @_;
    
    my @divisors=();
    my $start_num=1;
    
    if ($num == 1) {
        push @divisors, 1;
        return @divisors;
    }
    
    while(1) {
        if ($num % $start_num == 0) {
            push @divisors, $start_num;
        }
        
        $start_num+=1;
        
        if ($num == $start_num) {
            last;
        }
    }
    
    return @divisors;
}

my $target_num=10000;
my @result_list=();

while(1) {
    if(defined(first { $_ eq $target_num} @result_list)) {
        $target_num-=1;
        next;
    }
    
    print $target_num."\n";
    
    my @divisors=getDivisors($target_num);
    my $sum_of_divisors=sum @divisors;
    
    print "divisors=@divisors, sum_of_divisors=$sum_of_divisors \n";
    
    my @temp_divisors=getDivisors($sum_of_divisors);
    my $temp_sum_of_divisors=sum @temp_divisors;
    
    if ($target_num == $temp_sum_of_divisors and $target_num != $sum_of_divisors) {
        print "in result list";
        
        push @result_list, $target_num;
        push @result_list, $sum_of_divisors;
    }
    
    $target_num-=1;
    
    if ($target_num == 0) {
        last;
    }
}

print "@result_list";
my $r=sum @result_list;
print $r